src/Server/TextDocument.php

```diff
@@ -124,4 +126,25 @@ private function updateAst(string $uri, string $content)
             $this->asts[$uri] = $stmts;
         }
     }
+
+    /**
+     * The document formatting request is sent from the server to the client to format a whole document.
+     *
+     * @param TextDocumentIdentifier $textDocument The document to format
+     * @param FormattingOptions $options The format options
+     * @return TextEdit[]
+     */
+    public function formatting(TextDocumentIdentifier $textDocument, FormattingOptions $options)
+    {
+        $nodes = $this->asts[$textDocument->uri];
+        if (empty($nodes)){
+            return [];
+        }
+        $prettyPrinter = new PrettyPrinterStandard();
+        $edit = new TextEdit();
+        $edit->range = new Range(new Position(0, 0), new Position(PHP_INT_MAX, PHP_INT_MAX));
```

Comment:
I didn't found anything about it in protocol spec. Most probably it will depend on client implementation. It works for VS Code but if you feel uncomfortable with it I will reorganize code to store opened documents to calculate end position or traverse to last ast node and get end position from it.